chartpress.yaml

```diff
@@ -13,6 +13,7 @@ charts:
     # Dev: imagePrefix can be useful to override if you want to trial something
     #      locally developed in a remote k8s cluster.
     imagePrefix: jupyterhub/k8s-
+    baseVersion: "2.0.0-0.dev"
```

Comment:
It currently works as it should, because we have 2.0.0-beta.1 out and then the behavior without having a baseVersion set is matching what is correct.

```suggestion
    baseVersion: "2.0.0-beta.1"
```

Reply:
![image](https://user-images.githubusercontent.com/3837114/188829536-1b4fb6e9-38f1-46ec-8bce-720b7aa64d28.png)

Reply:
Hm. I was specifically trying to avoid needing to update baseVersion for every prerelease (i.e. only bump it on final release). The result would be that untagged dev releases would keep '0.dev' in between tagged betas. WDYT?

Reply:
So dev releases would always show up as `2.0.0-0.dev.git.5783.h70afc746` no matter if they are before or after a alpha/beta tag?

Not so happy about that, then we have version sorting issues once again, but between dev releases and pre-releases. I'll want to think further and read more in the PR you just opened in chartpress that I yet haven't looked at.

Reply:
> then we have version sorting issues once again

I think that depends. Should untagged builds be preferred to the latest prerelease tag? I think it could be a _good_ thing that the latest tag is preferred to the latest untagged commit.

One nice thing about the absolute commit count is you can see clearly the build order, regardless of the prerelease tag (humans, not machines).

I'm trying to find the right balance of maintenance tasks vs clear communication, and minimize the impact of this workflow change to solve the relatively minor issue of confusing dev tags mid-development. Bumping baseVersion during prereleases adds a manual, not-automated step to making prereleases (as opposed to just stable releases). Is the semantic ordering of `beta.1[.git.5700]` vs `dev.0.git.5701` important? If so, which is the best direction of preference? I'd argue that the ordering between the affected versions is _not_ important, and the best direction is ambiguous. This issue also only comes up _during the prerelease process_, not during normal development, where I think the distinction is of relatively little consequence, and I think reducing maintainer tasks has a larger benefit than perfect semantic ordering of a small number of necessarily similar unreleased commits.

But it is just one task per prerelease tag, so if you think perfect ordering is worth the extra step (and preferring dev to tagged beta is desirable), we can add bumping baseVersion to the prerelease steps as well.

Reply:
I'm having so many hats on now:
- Being a z2jh user and communicating to other z2jh users as a chart maintainer: I greatly appreciate if the intuition of version sorting is correctly aligning with how new something is. It can be relevant for example when viewing releases at artifacthub.io. People will get this by default I think, so anything besides this is a drawback I think.
  ![image](https://user-images.githubusercontent.com/3837114/188875834-0164625c-c2c1-42a6-b4a7-6a1dac3eb6c3.png)
- Being chartpress user and a z2jh maintainer: I appreciate if its a) easy to communicate what to do in RELEASE.md, and b) if what to do is little work. Optimizing for a) is far more important than b) in my mind - to run longer but in a straight line is to be preferred over running a shorter distance but needing to read a map first.
- Being a chartpress developer: I appreciate simple logic that is easy to document and maintain.

As a z2jh maintainer, I'd very much like to do the work to ensure dev releases are sorting correctly also in between pre-releases and not compromise on that - even if it means to update baseVersion often. As a chartpress maintainer I'd very much like to stick to relatively simple logic that is robust and easy to understand.

---

> I think reducing maintainer tasks has a larger benefit than perfect semantic ordering of a small number of necessarily similar unreleased commits.

I see this as a trade off between more work that is easier and than less work that is more complicated. I have too many hats on at the same time to think very clearly about things, but for z2jh specifically I'm opinionated towards not doing anything with baseVersion or doing it systematically so that our releases are always version sorting newer if they are newer.

Reply:
> I greatly appreciate if the intuition of version sorting is correctly aligning with how new something is. It can be relevant for example when viewing releases at artifacthub.io. People will get this by default I think, so anything besides this is a drawback I think.

I want to make sure we're clear that we don't have this now, which was the motivation for introducing the baseVersion option. Right now, _all_ dev versions are sorted incorrectly behind the last release, and (pre-2.0) inconsistently relative to each other. Only between a prerelease and a final release (where we are now) would ordering be mostly correct (but not entirely if more than 9 commits occurred), the rest of the time it would be wrong. As an example, in your screenshot that `1.1.3-n804` (and hundreds of other tags) should be between 1.2.0 and 2.0.0-beta.1, not below 1.1.3.

We've lived with "dev versions are not sortable except by humans looking at the `.n123`" field (or date) thus far. Every option with baseVersion is an improvement over the status quo in this regard. Without baseVersion, 2.0 is an improvement in that dev versions are _internally_ sortable, but still incorrect relative to stable releases.

The question right now is: _during_ the prerelease process, how should _tagged_ prereleases compare with _untagged_ dev versions. I genuinely think both answers are valid! Preferring a tagged prerelease to an untagged commit doesn't seem wrong to me. At the same time, during the beta process all commits probably are straightforward improvements over what came before.

It's a bit frustrating that the _only_ time behavior is correct without baseVersion (when the latest tag is a prerelease) is the time it's most tedious for baseVersion to behave correctly.


>  Optimizing for a) is far more important than b) in my mind - to run longer but in a straight line is to be preferred over running a shorter distance but needing to read a map first.

I definitely agree here. This is part of my main reason for omitting the baseVersion bump, but your thinking may be different! I thought it would be easier to make prereleases and not make mistakes if the only repo activity were 'push a tag'. But maybe it's more of a burden for stable releases to have _different_ steps from prereleases. WDYT?

WDYT about adopting tbump? I think any time making a release involves modifying file(s) and making a tag, having a tool do both atomically is valuable. That would mean baseVersion always matches the tag, and is updated just once.

It would _require_ https://github.com/jupyterhub/chartpress/pull/184 because without that, setting `baseVersion: 2.0.0` on the release tag would cause chartpress to fail.

Reply:
Actually, I was wrong about that last statement - we already accept `baseVersion: 2.0.0` _on the tagged commit itself_, but not after. So I think that's all okay already.

Reply:
Nope, sorry. I had https://github.com/jupyterhub/chartpress/pull/184 checked out when I tested locally. We do indeed need that patch in order to use tbump, which I think should probably be recommended in combination with baseVersion. It simplifies everything, _reduces_ the complexity of making a release relative to what we already have, _and_ ensures version ordering is always correct, even between prereleases.

Reply:
:heart: :tada: soo much thorough effort into this! Thank you @minrk I'm super happy about a robust tbump resolution, now use of baseVersion makes me confident. Its fine that the initial setup is somewhat tricky, especially now that chartpress links to an example setup.

Reply:
Thanks for all your review! I'm much happier with the combination with tbump as well.